pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -313,6 +313,15 @@ func (c *Controller) syncService(key string) error {
 		return err
 	}
 
+	if service.DeletionTimestamp != nil {
```

Comment:
do we need a similar fix for the endpoints controller?

Reply:
I am fairly new to this code, but it looks like in the case of endpoints controller we explicitly delete the endpoint instead of gc to kick in for the NotFound case. The check for DeletionTimestamp is nevertheless missing so we may end up creating new endpoints potentially similar to the original issue reported.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L372

Reply:
The Endpoints controller actually does not set an owner reference, so there's not really a parallel here. I think the Endpoins controller might predate owner references. There's an [issue from March 2015](https://github.com/kubernetes/kubernetes/issues/6877) that notes that Endpoints were never cleaned up by the Endpoints controller. This was [fixed in April 2015 by a PR that added a cleanup function](https://github.com/kubernetes/kubernetes/pull/7821) on controller startup. That [checkLeftoverEndpoints() function continues to exist](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L549) in a mostly unchanged state. 

That cleanup function does make an important exception for Endpoints with the `control-plane.alpha.kubernetes.io/leader` annotation though, not quite sure if Endpoints used for leader election ever have a corresponding Service though. Assuming they don't, this would be a non-issue for garbage collection.

So to summarize:
1. The Endpoints controller does not use owner references and will not conflict with garbage collection.
2. @abhat's link above shows that Endpoints are cleaned up within the syncService() function when a Service can not be found. 
3. We could potentially extend that cleanup logic to include a Service with a deletionTimestamp, but I'm not sure that would make a meaningful difference or be worth the effort.
4. It may be worth updating the Endpoints controller to use owner references.

Reply:
> The Endpoints controller actually does not set an owner reference, so there's not really a parallel here.

Ah, right.